src/dra/4A538.c

```diff
@@ -719,84 +719,81 @@ void ResetEntityArray(void) {
     }
 }
 
-#ifndef NON_MATCHING
-INCLUDE_ASM("dra/nonmatchings/4A538", RenderEntities);
-#else
 typedef struct {
-    /* 0x1F800034 */ u16 unk0;
-    /* 0x1F800036 */ s16 animSet;
-    /* 0x1F800038 */ s16 animCurFrame;
-    /* 0x1F80003A */ u16 w;
-    /* 0x1F80003C */ u16 h;
-    /* 0x1F80003E */ s16 x;
-    /* 0x1F800040 */ s16 y;
-    /* 0x1F800042 */ u16 xPivot;
-    /* 0x1F800044 */ u16 yPivot;
-    /* 0x1F800046 */ u16 unused12;
-    /* 0x1F800048 */ s32 index;
-    /* 0x1F80004C */ s32 unk18;
-    /* 0x1F800050 */ u32 unk1C;
-    /* 0x1F800054 */ s32 unk20;
-    /* 0x1F800058 */ s32 unk24;
-    /* 0x1F80005C */ OT_TYPE* ot;
-    /* 0x1F800060 */ POLY_GT4* poly;
-} EntitiesRenderer; // size=0x30
+    s16 unk0;
+    s16 animSet;
+    s16 animCurFrame;
+    u16 w;
+    u16 h;
+    s16 x;
+    s16 y;
+    s16 xPivot;
+    s16 yPivot;
+    u16 unused12;
+    s32 index;
+    s32 index2;
+    u32 spriteSheetIdx;
+    s32 facingLeft2; // set to 2 when facing left, 0 when not
+    s32 eDrawFlags;
+    u_long* ot;
+    POLY_GT4* poly;
+} EntitiesRenderer;
 
 void RenderEntities(void) {
 #ifdef VERSION_PC
-    POLY_GT4 _poly;
-    POLY_GT4* poly = &_poly;
-    EntitiesRenderer _r;
-    EntitiesRenderer* r = &_r;
-#else
-    POLY_GT4* poly = (POLY_GT4*)0x1F800000;
-    EntitiesRenderer* r = (EntitiesRenderer*)0x1F800034;
+    u8 sp[SP_LEN];
 #endif
-
+    s16 polyX;
+    s16 polyY;
+    s32 animFrameFlags;
     Entity* entity;
-    PlayerDraw* plDraw;
-    s32 temp_s2_2;
-    s16* var_t2;
-    s16 temp_a0_10;
-    s16 temp_a0_2;
-    s32 temp_v1_4;
-    s32 var_a0_2;
-    s16 var_a0;
-    u16 temp_a2;
-    s32 new_var;
-    u16 temp_v0_6;
-    u16 temp_v1_6;
-    u16 temp_v1_8;
-    u16 var_v0_3;
-    s32 var_v0_4;
-    s32 test;
-    u8 texEX;
-    u8 texSY;
-    u8 texEY;
-    u8 texSX;
-    u8* spriteData;
+    s16* animFrame; // might be a struct
+    u8 uvRight;
+    u8 uvTop;
+    s32 tempFlagCheck;
     s16** spriteBank;
-    u16 animSet;
-    u8** prestuff;
-    u8** stuff;
-
-    entity = g_Entities;
+    u8* spriteData;
+    PlayerDraw* plDraw;
+    u16 palette;
+    u16 animFrameIdx;
+    u8 uvLeft;
+    u8 uvBottom;
+
+    POLY_GT4* poly;
+    EntitiesRenderer* r;
+    poly = (POLY_GT4*)SPAD(0);
```

Comment:
This part is still needed to prevent crashes
```
#ifdef VERSION_PC
    POLY_GT4 _poly;
    POLY_GT4* poly = &_poly;
    EntitiesRenderer _r;
    EntitiesRenderer* r = &_r;
#else
```

Reply:
Okay, I think I fixed it properly then. I also moved the declaration of the `sp` array down, so that we only need one #ifdef for this. Let me know if it's still wrong :)

Reply:
This is untested but it's probably more like:

#ifdef VERSION_PC
    POLY_GT4 _poly;
    POLY_GT4* poly = &_poly;
    EntitiesRenderer _r;
    EntitiesRenderer* r = &_r;
#else
    POLY_GT4* poly;
    EntitiesRenderer* r;
    poly = (POLY_GT4*)SPAD(0);
#ifdef VERSION_PSP
    r = (EntitiesRenderer*)SPAD(14);
#else
    r = (EntitiesRenderer*)SPAD(13);
#endif
#endif

The issue is there's no scratchpad so you have to create some storage and use that instead.

Reply:
But we create the local emulated scratchpad with 
```#ifdef VERSION_PC
    u8 sp[SP_LEN];
```
, right?

Reply:
Ok if we're going that route then you don't need this part
```
POLY_GT4 _poly;
POLY_GT4* poly = &_poly;
EntitiesRenderer _r;
EntitiesRenderer* r = &_r;
```

Reply:
I think generally this way is cleaner since the size of these objects may not be the same, so hardcoded scratchpad offsets aren't going to work right. 
```
POLY_GT4 _poly;
POLY_GT4* poly = &_poly;
EntitiesRenderer _r;
EntitiesRenderer* r = &_r;
```

Reply:
Okay, changed PC to use this syntax. I also changed the two other versions to use `sizeof` in order to remove the ifdef, which is nice.

Reply:
> But we create the local emulated scratchpad with `u8 sp[SP_LEN];`

Unfortunately that is not very reliable when working with pointers. For example if you have `a = (void*)SP(0);` and `b = (void*)SP(4)`, on 64-bit builds a pointer is 8 bytes long, so now `b = SP(4)` will fall into an unexpected portion of memory.

This is why I came up with `void *_b; b = &_b` as it is more reliable than anything else on any kind of configuration.

Reply:
If we use `sizeof` with all `SP()` uses, would that fix things? That way even if a pointer is 8 bytes, the scratchpad will always align in a way that keeps everything in the right place.

Reply:
I think sizeof() fixes the issue, but what happens if the size of the structs exceeds SP_LEN? The `&_poly` method automatically allocates the right amount of space and you're not casting to u8 and back with is cleaner in my opinion. I think address sanitizer won't catch buffer overruns if you cast to `sp`. e.g. if the code did something like `poly->lastitem[too_big]` it would overwrite the next item in `sp` but not get caught

Reply:
Agree with Sozud. I was trying to use `sizeof()` on `RenderPrimitives` but it didn't work that well either unfortunately

Reply:
That makes sense. We'll keep this as-is then. PSP/PS1 use `sizeof()`, while PC will use `&_poly`.